File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
 ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
{
    ListNode *temp1 = headA, *temp2 = headB;
    set<ListNode *> st;
    while (temp1)
    {
        st.insert(temp1);
        temp1 = temp1->next;
    }
    while (temp2)
    {
        if (st.find(temp2) != st.end())
        {
            return temp2;
        }
        temp2 = temp2->next;
    }
    return nullptr;
}
};

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode i = headA;
        ListNode j = headB;
        while (i != j) {
            if (i == null) {
                i = headB;
            } else {
                i = i.next;
            }
            if (j == null) {
                j = headA;
            } else
                j = j.next;
        }
        return i;

    }
}